# Rosalind_solutions
# DNA: Counting DNA Nucleotides
Given: A DNA string s of length at most 1000 nt.

Return: Four integers (separated by spaces) counting the respective number of times that the symbols 'A', 'C', 'G', and 'T' occur in s.

dnaseq = 'GACGCGAGATTCCATCTCAATGTAATGCGCATGTGGAAAGACTACGAGACGGCTCCGGATGTGATTGAGAAGTTCAGCGTTCCACATGACCACCCGTATATGGGCGACGGCCCTCTGCATATGGCAGCGATCACCGGAAGCCAATAGATTAACTGCGCTTCGTGTCGTCCTCTTTACTATCTACGGGTGGGGCCTGAATAAAGACTTGAGTACTCGCTGGCCAGAGTTCTAGCGCCACCGTCTTAACTGTTCACCCGGGGTGGCACGACTACTTGTCTGGGCGTTTTGGTATACTTGGAGCCTCCTGAAACTGACTTTCGTTGGTCAACAGATGGAAGACCGAGGCTAACTTAGCAAGAGCTGGCCTGGTTTGGGATGACAACTCGAACCCCATGGCTGTGGTTCACATTACTAACTATAGTTCTTAGGTATTAGAGGGGAATCAAGATATATACGATATTAGAAGATTCAAAAGTCATTGGCCGCATTGAACATGAAAGTCGATGGACGCCAATCCGATCATCTAATACTGTGCCTGTCATACAGCGTTATGTCGATTAGACCCGGGGTAGGTCTTTGGACTAAATTGCATTATAAGATGAAAGATTATGTCGGCGCTCGTTGCCAATTGCATAGACCTGATACGTTGACCAAGGAATGGGAAGCGGCAGCTGGGGTATAAGAATGGCTTCGGCCTAATACTCCGCCGCCTTTCACCCAGCTCACAGGCTTCCGCTCAAGAACATGAGCTCTCGTCGGGATGCTTACCGAACATGCGTCGAGACCGGCACGACCGAGTCCAGTACTGCCGCCAAGATGTGGTGCGCAAGGAGGAGTCCCTACAGTCATGCGGGGCCAGTTCGCTATCGGGCACTCTACCCATCTTAAGTCTAAAGATTTTTGTCATTAACTGCTGGGGTGGATGTATAATATCTGGGTCTCCGCCCCCTCGTATGAGATAGACACATAGATCTGCGCTAGAGATCCAGTCATC'

def nt_count(dnaseq):

  A_count = dnaseq.count('A')
  
  C_count = dnaseq.count('C')
  
  G_count = dnaseq.count('G')
  
  T_count = dnaseq.count('T')
  
  return A_count, C_count, G_count, T_count

nt_count(dnaseq)

# RNA: Transcribing DNA Into RNA
Given: A DNA string t having length at most 1000 nt.

Return: The transcribed RNA string of t.

dnaseq = 
'CGTGGATTTCGGTGTACTTCGGATGCATAAAAGGTTCTGATCTCTCTCTGGCCTCCTAACCCACTAGTCCGCTCTGAGGTCATCGCAGTTATCATTACTTACTCGCTACTAAAGAGACCTACGAGGATTACTGAATAAGTAACTACCCCTGGGCGTGGTATCTTTCAGGTGAAGTAAACCGGGGCGCACCAGATTCGCACACGTACCACCTTACCGCATTCAGGTCTGGCGTCCCAGGGTTGCCCACTACTGCCCCAAGAGCAGACCTCAAACTTGTACGCTAACATGCGTTACGCCGATCTATCGGCCTATGCGGACGAGCTGGCGGTTTTCCGGATGTATCCACATGAGTCGGATTATGCGGACAACTGGTGGAACGCTGATTTGAGATTACACCGATTATCTACTGGATAGGAGTAGCCTCAGGAACCCTACGCGTTGTCCAGCTTAGCACCATGTCGGATCCATATATCACGAGAGCGCCATCGAGGTGCCCGAGAGCCGCCCCCTGCGCACTGCCAGCCCCTCCCCCTCGTGGCGATAGACAGGGCAGAGCATATTGTAAGGTCTTACTCAATCCGTACAATCTCTAAGGCGTCCAATGCTGCATGAGGGGGCTGAGCTCAGGAGGACCCACCTAAGTACGGTTTCTGATCCACACTCGCCAGTGCTTTTGGCGCGATCTAAGAGGTTGCAAAAAGGACCCTGCCGCGATGGTCATCCAATACTGTAGATGGTGTGTGAGTCAAATGTGATAGTGAAGCTGTTTGGAAGGACATACGGAGAGCAGTGCTATGCGACTATTCTACACCAACCATGGTAGACCAGTCTCCGTTCTCAGCCGGTATGCGTTAGCGGAGTGTGTATTGGGATGGAAGATTACTTCGTGACACCACCAGCTGTAATCACTT'

def transcribe(dnaseq):

  rna = dnaseq.replace('T', 'U')
  
  return rna

transcribe(dnaseq)

# REVC: Complementing a Strand of DNA

Given: A DNA string s of length at most 1000 bp.

Return: The reverse complement sc of s.

dnaseq = 
'CTGTTGTATCAGAAAGTGCCGTCTGTAATGGTTGTCCCGTCTCTTCTAACCAACTACAAAGTCAATGCTTTTTGGTTGCTAGCGGGCGTCGACAGGAACCCTATTATATGCGATACGGCTCGTGATGCAAAGCGCGCCCATGGGTAGGACCAGTACATCGGACCGATTCGCACCAGGCCCGTCCGATAGACTCGGCATCATCACAGTAGTTTCATGCCGGGCCATGTCAGGGGACCACTATCTGATTTACCGTTCTTTTCGCTCGTAATAGTTCGGTACTATGTCTTGGGGCAATCGAGGTGCTTGTATATGTCCCAAGACTTCTCGGCTCTACACAGACGAATCGTCTGATGTGGACTAAGACAAAGTAGCTTCCTGATGCGTCAAACAACGGGGCCTTAGTGTGTCACCAGGCAGAAGTTGTATCGACCAATTATTTGTTCGATTTAGCAACAGGTCCACCGTGTCAGTGGCTTACCATTCCTGAGCCCCACAGGGTACAGGGGATTTGTTTGCCACCCGTCGGATCAACTGAACATCAATGCGTCTGTGCTCTCAAGGGATCCTACACTTTTATGTGCCCTCGAACCTGGTCTAGTTATTTCCGTGTACAATTTGAGCCCCACACGGTCTTCTGGTGTAATCTGACCCACGTAATTCAAGAAGAGGGAAACCGAAACAGCGGAAAAGATTCTGTGCGTTTGATATGGAGCCAAGACAAGCCCAGGGCGACCAATTGTCAGCACGCGGTTTCACAGAACCGGCTGATTAATATTCGGGCCCTAACCCGGCCGGTTTCATAGTTCTGTCCTCGCGTGACGTATGTTCGCTGTGCTTCGGACGCAGAGGGCGTCGAAGCGGACAACACGAGTACATCTCCTGACGACACCTTAC'

def complement(dnaseq):

  bp_dict = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}
  
  rev_dna = dnaseq[::-1]
  
  comp = ''
  
  for nt in rev_dna: 
  
    comp += bp_dict[nt]
    
  return comp

complement(dnaseq)

# GC: Computing GC Content

Given: At most 10 DNA strings in FASTA format (of length at most 1 kbp each).

Return: The ID of the string having the highest GC-content, followed by the GC-content of that string. Rosalind allows for a default error of 0.001 in all decimal answers unless otherwise stated.

Rosalind_1900 = 
'GAGGCGGCCTTGACGCTAGAACAAGAGAACAATTAGAGTGCCGGATGCGCTCTGGGTGGGCGTGCTACGATGCAGAATATTGTAGTTGACGAAGAATTCATGATGCAATTCCGCTACCCCGTCCGAACGAGTGCCGCTGGACTACCAGGAAGAATGGGATGGGGGCTAGGTACATCTCAATCATCTGTATCAATGTATATTCGTCCTGGACTTAAATCAGCTCCTATGAGTGGGGGGGCAGGCAGGACCAGCGTTCATTGCGTACTGCCGCTATCGTGGGCGCATAGGCTGCATGGTCTAGTCCCTCCCCGAATAATATGTGTCGCTGTGAGTCTTTTCATTTTAGCCAGCGTGAAAAGCAAGCGTTCCACGTTCGCGAAGCTGTGGCAATTCAAGTGACTGTGAGACGGACTCTGGGGCATCGAGCAGTCTCCGCTAGTGTTGCCTAAGGCAGTAAGAACTAGTCAGGGAAATTTTTGTCGCAAATCAGTACATAACGGGTCATCATCTGACCGACCTCGCAAAAGGTGTAACGCACGACGGAGACCGAACTGCGAGACTAGAAGTGCGTCTAGCTCTGAAATACAAGGATTGGTACAGTTGAAGGCATGGTACTACACTATGTAGTGCTGATGGCTCCATGATTCAGGTAATAAGCGCTTCTGCTTAGGCGCAGAAGAAGCTTCAGTAAACGGCGGGATTGCAGGGCCCAAGCTGTCGTTCTACCTTCCGACCTATCAACCTTTTATGCATATGTGTAATTTACTGCTCATTCGGTTTTTAAAAATCCGAGCGGTTTGCCGCGGTGACAAGAACAGCGCATAACGACGTCGGGATACTTCTAGCTATAAAACATGACCATGTCATAGCGTGCGCACAAGTTAAGCAGATACCGTGCCGTGGC'

def GC_content(seq):

  GC_content = ((seq.count('G') + seq.count('C')) / len(seq)) * 100
  
  return GC_content

GC_content(Rosalind_1900)

# HAMM: Counting Point Mutations

Given: Two DNA strings s and t of equal length (not exceeding 1 kbp).

Return: The Hamming distance dH(s,t).

seq1 = 
'ACAACAGGACACAAACCCCTCACACGTGTAACACATTTCGCCCCATCACGCGCCACACATTTCCCGGTAAATCGTCTGATATCCAGTTGAGGCTCCACGATATGTCCGACAGACCATCTCTCCAAATTTGCGGTACCAGATTCTGCTGAAAGGGGCTAACGATCGTATAGCAATTTACCTAGTCTTGCTTCACTCACAACGTCATAAACCAGGTGTGATGTACTGCTCCGCCCAAAAGTAACCCGAGATACATTCAACGTTCCCTACGAAAAGGTTATATGGTATAATGAGCTGACCGCCGACGTAAGCATGGCTCTCCCTTCAGAAAACAGTCGGCGGAACGGCGCCTCTTGGCACGCATTCTTTCGACGGCTAAACCCACTAGTGAGAAGTACCACTATGTCGTTCTGTGGGTAAGACCCGTTGAGCGGAAACAGTGTAAGTTGGCAACCAGATCTCCCGATAGGTGTTAGTAATCCCTGCAACTCGGTTGCTCACTCACCGCGCTCCTCTGCAATAGAAACCGTTTGCTTGAGACCTTAGACTTGACTCCGTAGTAAGCCCTAACTACACGTCTTAATTAATTTTAGCAACGATTCCAATCAGCGGAAACATTTAGACACCAATCTACCCGGACTTCATGCATCATGGTCAGAGTTCCACATCGTCGTAGTTCCACCTCACTGCTCCGCACGGTTTGGGACTTTCGCACCGATCCTAGACTGCCCAATGCAGCTCTTTAGCCAGCTACTGCTGAGGGCCAAGGACTGATACCTAGTACCCATAGTTGGGGTTGCCCGCTACCAACGAAACAGAGTGGCGGCCCAGCAAAACTAATGATCACATCTGCAGATGTATTCGTCTTTAAGCTCCGGATGTTGGCTCCTCGGGCCAATCTATCGAGATCAGCAAGTTCTGAGCGATAAAAAAGATCGGCCCTACGCCCTACTGTAGACTTG'

seq2 = 
'TTACGAGTACCTTTAGTTCCGACTATGGAAATGGAATGCACCCTAGCACAAGCTAAACTTTTGTCGTCAGTCCCTCTTGCACCCTGCAAAGGTTCCTCGAGCTCTACTCCTATCGAAAGCCTCGAACCTTCAGTAGCAGCTGTCGCTTGATGGCGCTCTCAAAGATACGCCGATTAGAAAGATTCATTTTTACTTATGGAGTTGGCAATCAGGAGTCTGGTGGGGAGCCTACCAAAAGACCGGTCAAAGACACGGCTCGCAGGCTTAAATAAGGATACGTGGTAAAATTTGTCGCTCCACTGAGGTAGCGGGGCTCTTACTCCAGATGCAAGTTTGCGAACAGCCGCTGCTTTGGCATGTAGAGGTTGGCGACTAGGCCCACCAGTGCGCATATACACACTGCCAACCTCCGCTCAACACCCGTCGGGCTTGAACAGCTCGTAATGTACGCCCGCTAACAATTATATTGTGAGCAAACCCTGCATCACGCCAGCACAGCGAGCCCGGTCCTATAAATTAGAAACGGGGTCGTTATGACCTTAGTCTGGACCCAGTTGTCTGCAGGAACCTCACGCATAAACTTTATCTAGCTAGGAACTCGATCAGCGTTAACATGTTGAAACCTAACTTGCAAGACGTTTTACTTCCACTTCCTATTAGCACAGGTTTGGGTTTCCTCCTCTCAAAGCCGGGTCTGTGGCTGCTCGCCCGCTAAGACTGGAGACTGGAGGGCTGACCGTGAACCACTACCCTCTTTAGGAACTAAGGCGCTTCAACGATCCAATTGTAGGAGTTCCCCGGTGATAACTGAAGCTAGACGCGAGCCATCATACTTAAAGGCCGAATCTCTAGCTATCTTAGTCGCGATCAACTTATTCACGGGTATGCCTAACAACCTAGCCAGACATGAAAAGTCTCTGATATTATCTAGAGTGCCAGTCCGTCCTCCTTTAGAAAAG'

def dH(seq1, seq2):

  mutations = 0 
  
  for i in range(len(seq1)):
  
    if seq1[i] != seq2[i]:
      mutations += 1
      
  return mutations

dH(seq1, seq2)

# PROT: Translating RNA into Protein

Given: An RNA string s corresponding to a strand of mRNA (of length at most 10 kbp).

Return: The protein string encoded by s.

protein_dict = {
    "UUU":"F", "UUC":"F", "UUA":"L", "UUG":"L",
    "UCU":"S", "UCC":"S", "UCA":"S", "UCG":"S",
    "UAU":"Y", "UAC":"Y", "UAA":"Stop", "UAG":"Stop",
    "UGU":"C", "UGC":"C", "UGA":"Stop", "UGG":"W",

    "CUU":"L", "CUC":"L", "CUA":"L", "CUG":"L",
    "CCU":"P", "CCC":"P", "CCA":"P", "CCG":"P",
    "CAU":"H", "CAC":"H", "CAA":"Q", "CAG":"Q",
    "CGU":"R", "CGC":"R", "CGA":"R", "CGG":"R",

    "AUU":"I", "AUC":"I", "AUA":"I", "AUG":"M",
    "ACU":"T", "ACC":"T", "ACA":"T", "ACG":"T",
    "AAU":"N", "AAC":"N", "AAA":"K", "AAG":"K",
    "AGU":"S", "AGC":"S", "AGA":"R", "AGG":"R",

    "GUU":"V", "GUC":"V", "GUA":"V", "GUG":"V",
    "GCU":"A", "GCC":"A", "GCA":"A", "GCG":"A",
    "GAU":"D", "GAC":"D", "GAA":"E", "GAG":"E",
    "GGU":"G", "GGC":"G", "GGA":"G", "GGG":"G"
}

mRNA = 
'AUGUUAAAGCGAUUGGUCACCGCUCCCCCGCCAAUUCUCAGUCCGCCCAUGGUGUGCUGCUUCUUCUUAACGUUUGCCAACUUCCUCUCGGGUCGCGCUUCACAGGGAAUGAAUAGUGGAAUAUCAUUUGCUGGACCUCCUCUGCACUCCCGGUGGCGGUUGCGUGUUAUAAAUGGUACAUACAAUGAACGUAUAUCAGUUACUGGCAACCGUCGGCGCAACCAAGCACGUGCUUCAUUCGGUCCUGAUCUCCAACUAUGGGUCGUGCCGGUGGCGCCACGUAUUCGCACUCGGGCAGCCCGAGUUCCCUCUGUUAUUAUUACUAAGAUACGCCGUAAGAGGGCAAGCGAAGGGCACGCCGGUUUUGAUCCGCCACCACUACGAAGGCGGGGCGAAGGCAGACAUUCUCAUAGUUUGGAGUUUAGCGACACAGUUCAUCCCGGUGUCUCUACCAUGACGGGUUGCGAACGCGUAGUUCGUCUCGUAGCAAGGAAAGGCGUAUGUUCUCGACUUGAGGGGGGUAGAUCGCUGGCAACAUUGACAUAUUGUGAUAAUAUUAAAUCAUACUCAGUAGAGCGAAUAUUAGGCGAAGUUGUUAGAUGUCCUGCCCUAUGGUCGUUUUGCACGCCACGUUCGUCGCUUGGCCCGCCGAUGACGUUAAGCCUGAAGCAGCGAGGUGAGAGCUCUCGCCGUCUUGUGUCAUUAAGCCAAGGAUGUCUCUGGCCGUCAGUGCGCCUUCAGAGCCAUUGUUGCGAAUUCUGCAGGCAAGUCGAGGAGACUGCGUCACGUAGGUUGUCCGGCUAUACGACCAAUACCGAAUUUAGUCGUUUAGGCAGCCCAGGGAGCAUCGAAGCGUCGUACGGGCGGGCGAUUAGACUACCAAAUCCCCGCGGAAUGGGCUUACUCAUCAGUCCAUAUUGUCUUCUGCUACGUAAGCCGAUCCGGUGCCGCAUAGACGUCACUAAUUCAGUCGUGCUAGCUAGAGCGGCGGUACUCCCGGCCCUGUUUAAAGGUUCGGAGCAUGUGAGGCGGCCAGGCGCAAAGCGCGCCUUGUUGUGUAAUCUGGCGCGGGAUUUUGGUGGGUUUACGGCAAGUCUGUUGUCCUGGCGUACAAGGGUAAAGGUCUGUCCACAUAGCGUACACUUGAGAGAGCAAGUUCGGAUGCUGCUUAUUGGUAUCCUAUCAAUUGAUCAAACUACUCACUCGUCGCUCGAAGUUACCCUUCCCUCCCGAGCACACGGUCGGGAUCAACGACUAUAUAUUGAUUUCCUUACUGGACAUCGGCACGAUGCUCCGUUAAUUACUCAGUCAUCUAUUAGGCCAGGCUUGAUUCCACGAUCACAUUACGGACGCAGUACUGUGACGUACAACCGUACCAACCCACAUAAACAUCUAAAUCUAAGACGGGUCGUUUCUCCCUCGGGGGUUCAGCGAGAUGAUUUGGCGUCCUAUUCCUUCCGUCUCGCAGCCAGUCGGGCUAACAGAAUCAGUGUGAUCAAGAAGUCGUUUGGGCAAACUGUGGGCGUCCGCUAUUCGAGCGAUAGGUUCGUGCCACUUGAUCCUUUAAGGGAGUCCGUACGGCGUUUUAGAGCGAGUUGGACAAGUCCGAGGGACCAUACUAAUUCUCCGCUAUUUAGCGAAGCUCGCAGCCUCUUCAAUGCGACGGAUGGAGAACCGAGACUACCCACCCUGAGCCAUGAUACGGUAUGCUCAUGCAGCGUCAAGCCUAGGGCGUCUUAUUAUCGGCGGUGUCGUCUCUCAGCGCUAGCGCCUGAGUCCUCAAAUACCCCCCGACUGCUGAAUCAGUGUCCUUUCAUGCCGGCUAAAAUAGAGCUAUAUCUUUCUCUCUAUCGCAUUCCCUACAUUGUAGUCUCACGCAUCUACGCUGUGCACUUGCGGGUACUUAAACUGAAACUAAUGCUUCGUGCCACGAAAGUGCGACGUUUGGGCCGCUGCUCAUCGCACGAAAUUGAAUCGAGUUUGCAUGAGGUGAGGGUACCCCGACCCCGACGCCUCACAAUAAGUCCGAUUGUCUUCAAAUCCAUACAAAUAUCCGUUGAGUGCAAUGGGCAUAUGCAUCGUGUUAAAUAUGAGUCCCGUGAAAUUUGGGCAUGUACAUGGACGCCAAAGAGUCGAGGGUCCGCCAGUGAAGUCUGGACGUCUCGGGGCAGGAACAGCGGGGGUUGCAUGCCGAUCUCGGGCACAACAGGUGGCAUAGCCGAGGCAAAGCAGAAUGCGUCAACUAAGGAGAGUGAAAUGCAACAGGACGGAUUAUCUGAUAGAGCCGUACAUGGAUGUGGCCCAUGUGUCCCAACAGAGGGCCAGGUUUGUUAUUCUCGCAUUUCGGUCCAUAUAGGUCUCCGCGGGUCUUUAGCCCCCUAUCACGUCUUUACGAAAAGGUAUACUCAGUGCUCAUUCCCCAACUUGAGGUUAGCAUGUUAUCUUCAAAGCGAAGCCAAUCAGUCAGAUAACGUAACCUUGCCAGUGAGAAUGAUCUCCCCCGGGACCGUAAGACAAAGUCUAGUCUACACAGGUACCGCUCAAAACCCAUGGGCGCAAAUUGGUACUACGUGGUGGUCUUCCCGGUCGGGGCAAUAUAUACUUGGAUCGGCCCCUAAACAACCCUGGUCGGUAGCAGCCACUACAAGCUGGUACCUAAGCAGUAACGAUUGUCUGAUUAGAAGGGGACGUGUCGAGCCGAUUGCUCCACACAGUCGUGUAGCCGAUUUCAUCCCGAUUAAAUCGCUUAGAGUAUCGCAUAGUCGGACAUACGGCCGUAGAGGCCGUCGGAUGAACCUGUGGCAAACACGUGCAGAUCCUAUAGCUCUCUCCCUCGGUAACAUGACACAGGAAAGUAUGAAUACUGAAUUGCUACCCCCCCCUGCAGGCUCUUUGUUAACUACGGUUCGUGCGGAGUCUUGUGCACGCGUAAAACCAGCGGCUCAGCUAUUAAUGAUGCUCAUUAGUUGUAACAAAAGCAGCAUCGCGAGCGGCGUUUUUGUUUUUAUGAUCCUGCGAACGACCUCAGGCAUGCACGCGCCCGGUCCUCCAGCACUUAAAUAUUGUAACCGUUCGCCUACGCAGACAAACGGAGUUGCAUUUCUAGCAUGGAAACGAUUGAAAAGAGAAAAGGGUGGGUAUAGCUCCCUAGUCGGAACGUUUCAAGAUAUAGUGGUCUUACUUUUUGCAGGGCAUAGUGUAUCGCACAAGCCUACUUCUGCUAUCUUGGGCUAUCCCGUUUUAUUUUCUUGGGCUACGAUGGUCCAUAUUCUGCUUAGCCCUCCCUCGCACGCUAGUGUAACAAGUGCUGGCGACCUGUAUACACAAAUGGUGGGGCGGGCGUCCGGUUACCACAAAGUAACUACUCAGACUCUACGAUUCUCACUCACCAUACAUUCCCCCUACACUCAGGAGAUCUUGCGUUUCUUGGGAAGUAAUACGCCGAGUACAAUAGACCAGCAGCGAAACCCACGUAGGUCAGCUUGUCCGAAAGCGUACCCUAUUCUGGCGCUAAGUCUUUCACAAGUCUUUUCCGGGGUAGCGAAUAUGUCUAGAUACCGCCGAAUGUUCUGGAAAACAUGCGACAGUCAGCCCGUGAAGAGCGAGGCCUUCCAUUCCCGAUUUCACGUGCCUACCACAUUGCGACUGCCAGAGGGACUCGGGAUCCUCUCUAAAAUCGCUACACGCUUGGUACCCGCCAAAAGUAUCCCGUGGCGCGGAUUGAGGGUAUUGUUGAGCCGUGCAUGGUGCUCCGGACAGAUCGCGAGUCAUACAUUGUUGAUACACUCGGGUAAUGGGCUAACACAUAACAGCGGUACGACUAUUGCCUUUACUUACGCACCCGUUAUGACGUAUAACAUGAUGUUUUCCGAACAUCAGGCCUUAGACAAUGACAAAUUCAUGGGCCCUUUUUCCGGGUAUGUUCUCCGGCGCAUACGCGCCAGGGCAGCAGACCUAUACGCUCAAUCAUUAGAAGUAGAAACGGAACAAUGGUCAGCUUAUUGCUGGACCAGUUUGGCAACCGAAUUUGAACAGCCCGCACAGGCCGGCAUAGUUACGUUGCCUCUCACAGACUGUUUCCUACAUCAAUAUGACUUGCGGUCCAGACGGCAUAUAUACGCUGUGCUUGUGCCCACCAGACAUCUACGGCCUGGGAGAUCAGAUACCCAGCAGAAUUCUGUGUUGAUGCGCAUGAUAUCCAUUACCGAAACCGUUUACGCAGUGACGAGGUCAUUACCUGUAAGGCUCUCGAAAAAUGCCUUGGAUACGGGGUUCCGUCUGAGUAUGUCCCUUAUAGGUAAAGGGCUGUUUGUUCCGUCCUCGGCACCGGUGGGCCCACUAAGUGCAGCGCUCUUAACGCAGAUCCGGGCAAACAUAGUCCCUAACGCGGGGGGAUUUAUUUGCAGCUACAGAUACAGCGUUCUGGGAGCCAACAUCUGUCAUGUCAGUCCUGGACCCUCAUUAAGGCACAAACAGGUAGCUUUCCUAGUAGCUAAGUUAGACGCCAAUGAUACCGCCCCCUUAGUACAAAUUACAUGUGCUAAUGCUUAUUUGGCAGUCAGCGAGCUUCAAGCCCUCUUGCAUGUCUACGUUUUGGAUUACCUGAAAACGGACGUCGGCCCGCGAUUAUGUUCGAUAGGAGUCCCUAGACCUCUAAUUCCACGUAUAAAGGUAGCUGCACAUCUACUUUCAGUGUUGGUAUUAUAUAGGCCUUUGCAGAGCGCGACACGUAACUUACACAGUCACGGGAGGGGACCAGACUCAAGUAACUCGCACGUAGGUCUGCAUGGGACCGCGUUCUACUGGCAGGCACACACAGCGAACUUUGUGACGGCUAGAAACGCUUGGGACCCCCAGGUCGACCUUGGCGCAAACGUGGGGCACCCAAGUUUUGGCCGUUAUAGUGACUAUCAGACUGUCCGUUGGCGUAAGCAGGCCUUGUGGGUAUGUGGCUUCUAUGAUGCGCCAAGGUUUCCUCGUGCUUCGCAGCUUCCCGCCUGCUAUGUACGAUCAGUCCUCGCGAUAUGUAUUGACCAUUCUCGUUACAUACGGAUAAGGAGGAACUCGGCCGUUUUUGAGUUUCCACAGAUCGACCAAAUACUGCAGUCAGGAUUCGACUGGUGUGGGCGUCGUCGAUGUAGCGCUGCCGAAACCAUUCUUGGGAUAGCCCAAGUCAAGUUAUUGGCAGUACCGGAUCGUGAACGGACGUCGCCCGAGCAAGGGCUCCGUACCAAUGUUAUUGGCCGUAUGACGAUUUACCGCGACUCUCUCAGCAGCGUCUUGAACCGUCUGCGUAUGAAAGCAAAGCACUCGUGUAGAGGGACGCGUGGGACGGGUAUCCAUGUCCACGGGGUUUUCCUGGCGCCCUGCCUACGCCUGGCGGGGUUUGGUUUGUCCCGUGUUACCCAGGCUCGUCACGAAAGGGUCCAAACGGCGUUAGUACGUUCAGGGGACCAGAAUGCAGCGGUAACAGCCUCCCUCAGCAAAGCGCAAGGCUGGAGAGAUGUCCCUCAAAUGAUGUCCUUCUACGACACAUGUGCCGAGGUCAACCAGAUACCUAGAUCCAUUUUAUUGAGCUCGAGAAAGCGUGCUAGUGAGCCUUCUCCUGUCCGGCCUGUUGUAUGGCUCAGUCGAAUUAAGUUCGCUUUUAGUGCCAACUGCGGCGCAUAUGAGAUUUAUAAAUUAGCGGCAGAUUGGGGCUAUCGGAAGAGUGUUGUGGGAAGCCAGAGAGUUGGAAUACACAACUGGGCCUCUUCACCACGUCCCUUUACGUUCGUCCCGGUAAUGGUUAUGUGCCUUGUGGUGGGUGAUCCAUACCCUAGCUUAUCGGGGGACGACUUGUGGAUGCGGCCGCGCAACAAGUCUUCGGAAUCCCAUGUCCGGCAAGGCGGCGAUGACGGUACCUUGAACCCCGUCGCGAACGCAAUGGCAGUUUCAGUCGAACUUUUGGAAGACGGAAGGGGGAUGCGGCCGGUGAUUCCAUGCGAUAUGUACGUGUUGCUUCAAAAGCUCCGACAGCAUAAGUCCGGCGCCACAUUCGGAGGAUUCGGCUACUGUAGUAAGGGAACACCGUGGUUGUUUCCAGAUGAGAUAGGGGAAAUCUCAGUGAGGCACGCUAACGGCGGGAGCCCACCUCAAAAUUCACAAAGUUCGUUGUUGCUUAGGCGAGUAACCAAAGCACUCGUCUCGCUCCUAUUGUGCGACUCGAAUGGAACAGCGUGUCUUUCCUUUCCGGGGGUCCUCCCCAUAUCCACCUUUGUUAUCGGUGCUCUAUAUAACAUACCCUGGUCGCAUCGCUGCCGCAAGUCAGGGAGGUACUCUUGGAUUGUCAAGACUAAGGGGUCCCCCGCGGUACCAUGUGGAAGCAUGAAGCCUCCUGGGGUGAUGUACCUAUUUCUCAGGUUUGACUAUCAAUUCAACAUGGAACAGGUGGGGAGAGAGGGCACACAUGGUCUUACACCUCGUUGCCGUAGGAAAACAUCUUUUACUACAGAGGCUAAGGAACCAGGCUGUUUGGUGUUAAAGGUCCUUAUGAGGAACCGCUUAACUAUACAAAAGUUCGUGUCGGAAUUGUCAGUGGCAUGGAGCGGUUCCUCACAUUUCUCUCUUCUCAUACACCCCGAGGGCAUACCACGUAGGAAAAACGAUACAAACUCUAUGCACCGGUAUUAUUAUAUUCGUUGUCCUUGGAUAAUGAAUCGAGUCUGUCGAGGCCCUACCAUCGCUCAACCAUCGAUUCUGGUGCUUCUUCAACGAAGACAAAUGAAGGACUUUCGCAGACAGAUAAAACAUCCACAUACUGUAACAAUCGUUCAGUACGAUUCGCGAACUGUUGAUUGUGGCCACCGGUAUAACUUGAACCCGGUCAAUUCACUCCUUCGCGUCUACUACAGUCUGGCGAUGGUGUACCGUUGGAUUAAAUUGUGCCGGCGUUGUCUACUUGAUCUUUUGGGCUCGGCGCUUAUGGGUCAAAGGUAUCCUAACAGUUGGAUACUUUCGCGUGAGGAAACAAUUUUUGUUCGUUGCGUAAGCCUCACUCGGUCCGGACGCAUCCGUCCAAGUACGCAUUUGUGCCUAUGGGGGUGGCGACCUGGUUCUAGGAAUGAGGAGGCGGAGGGCCUGUUAUUCCUGUAUGGUACACCAAGAAUAACGGUUCAACCUGCCUGGCUCUGUAUAGCGCUCUGUCUCCGGGUGCGGAAGCAGUGCCUGCUACCUGGGACAAUUAGCUACAAUCCAUUGAUUCUUCGCCAUGCUCCGGCACGUACAAUAAGCAUCGGUGAAUGCGGUCACGGGCUACAAACCACGCCCAUCUACCACUUAGGUCUUCCAAUUAGGCGUAGUUCUUAUGGAGUUAAUCCUGUGCAACGGGCGUGCGCAGGCUUACUUGCAGAUGAGUCAACUACUAGCUUCCUAGGUACAAUACCAUUGCGGGUCACUCUUCACUACGUCAACGGUUUGGAUAGAUACAGGGAGCUAUCUAAGUUUGCACGGAGCACACAUAGUCGGGCCCUCGCCUUCAAUAUACACUCUCCCCGCUGUAGGUCAGGCAGAGCUCACAUUCGGAUUCUCCAGCAUUCGAAUAUUGUCGCAGCCUCGAUUUUAAACUAUUUGGGCUGUAAUCUCACAAUGGUUCAAUUACAUGUUCAAUCAACUGUGCACACAUCGACGAAAUCUCUUCCCAGUCUUCCCAUAGGUUGGAUACGUGCCGUUGACUCAACAAGUACUCCAUCUUCCGACCGGCGUAAAGGUGGACGGACUAAAUUUGCGAACCUGGGGAAAUUUUUUAUUUGUCGGUAUCGCCGAAUACCAGAACCUCAUUGUCUGGGUACAGGUCCCACGCAGGGUUUACAGAAGCCUGCGUGCUGGUCAUGCGGUGCGUGGUGUAGAGGUACAAUUCCCUGGAAGACUAGUUCAAGGGAAGAGAAGGUGAACGCUAUCCGAAGCGGCUCGGGUCGUACGAAGCGCUCUAUUCUCGGGAUGCGGCACCAACUUUGGUAUAGGUCGAAUCGCCCUCCUGUCGCCGCAUCGGGUUUGGAGCCUAUGUACCUGCAGGUUUCGCCUGAGUUAAUGACCCCAGGCCACAGGAACACGAAUGGCUUCGAAGUAAAUCAGUUUGUAGCGCCGCGAGCAUCCAGACUUGUGGAGGUUAAUUCACUCUUCAACGCAGUUGAUCGUUUAUGCACGGUUUAUGAUUUUGCAAAGGAUCCGAGGCACACCAUUAACGCGCUCGCUCUCCCUUGCGUUUCAAAGUACCCUGACAAACAGGCAUGGCUACAGUGUUACUCACGGUUUCACGGCCGAUCAAAUGAUCCACCAAUCGCGUGCUAUAUUGAUGGCGUGAGGGCCUGUGAGACCUUAAGAAAAAUCCCGUCCUACCCGCUCUAUGUUUCAGAUGCGUCAUCUAUCGACCGGUUGACGCUUACGGCCCGGUGCUCCAGUAGGGCUAGGUCCUGCAGCCAAUCGCUGCGUCCGUUCGUGCAUCGGUACUGGUGUGACGAAAGAGACCUGAAGCUCAAGACUUUUCUGGACGUAGUGGCUGAUAAUCCCUCAUCGUGUUUCUUCACAGUCUUUCUGUUUGCUACUCCCAAGCGAUUGCGACGUGAGUGUCCGGUCCCGACUGUAUUAAAAAUACAUUUGGCUGUUCAGUUCAUAGGCCAGUCCGUAGGCACGGGGGUCAUAGAUGUUCCAUGA'

def protein (seq):

  protein = ''
  
  start = seq.find('AUG')
  
  for i in range(start, len(seq), 3):
  
    codon = seq[i:i+3]
    if codon in protein_dict:
      aa = protein_dict[codon]
      if aa == 'Stop':
        break
      protein += aa
    else:
      break
      
  return protein

protein(mRNA)

# SUBS: Finding a Motif in DNA

Given: Two DNA strings s and t (each of length at most 1 kbp).

Return: All locations of t as a substring of s.

seq_s = 
'GCGTCTTATAGTCTTATTAATGAGTCTTATTCGAGTCTTATGTCTTATAATCTGTCTTATTTAGTCGTCTTATACGTCTTATGTCTTGGTCTTATTTAACCGCGTCTTATTAGTCTTATGTCTTATGCGCAGTCTTATCGTCTTATGAGTCTTATGTCTTATCACTCCCTGTCTTATTTGTCTTATCGTCTTATCGTTAATTGGTCTTATGACGCCGGAGTCTTATTGTCTTATGTCTTATGTCTTATGTCTTATTGTCTTATGGGTCTTATAGTCTTATGTCTTATGTCTTATTCACCAACCATTTCGTCTTATTGTCTTATCCGGTCTTATTCGTCTTATCGTGTCTTATCAGTCTTATAGGGGCAAGTCTTATGATAAGTGGACCCTCTGCGAAGTCTTATAAGTCTTATACTAGTCTTATGCGTCTTATTGGTCTTATGTCTTATCCATCGTTTACGTCTTATTGGTCTTATCTCTGTCTTATGCCGTCTTATAAGGGTCTTATATGTCTTATTCGTCTTATGTCTTATGCTCCGGTCTTATGCCGTCTTATCGGTCTTATTATAGTCTTATGGTCTTATAAGTTAGTCTTATGCTGTCTTATGTCTTATAGAGTGTCTTATGGAGTCTTATCGAAAGTAATTGTCTTATGGTTTTGTCTTATGTCTTATGTGTCTTATCTGTCTTATACCCTCTGTCTTATGTCTTATCGGTCTTATAAGGTCTTATGGATTAGTCTTATGGGGTCTTATCTTCTTGAAGTCTTATGAGTCTTATTAAGTAAACGAGTCTTATGTCTTATAGTCTTATGTCTTATCTTCAAGTCTTATTT'
seq_t = 'GTCTTATGT'

def motif(seq_s, seq_t):
  locations = []
  for i in range(len(seq_s) - len(seq_t)):
    if seq_s[i:i+len(seq_t)] == seq_t:
      locations.append(i+1)
  return locations

motif(seq_s, seq_t)

# PRTM: Calculating Protein Mass

Given: A protein string P of length at most 1000 aa. 

Return: The total weight of P. Consult the monoisotopic mass table.

protein_mass_dict = {
    'A': 71.03711, 
    'C': 103.00919, 
    'D': 115.02694, 
    'E': 129.04,
    'F': 147.068,
    'G': 57.02146, 
    'H': 137.05891, 
    'I': 113.08,
    'K': 128.09496, 
    'L': 113.08, 
    'M': 131.04,
    'N': 114.04293, 
    'P': 97.05276, 
    'Q': 128.05858, 
    'R': 156,
    'S': 87.03203, 
    'T': 101.04768, 
    'V': 99.06841, 
    'W': 186.07931,
    'Y': 163.06333
}

protein = 
'EHKMKHHPQEDLMRPSAIFQKIEFMFFRRPYAMCDYCLHPMHPFEFYENFRYTYCVNRDQVHFQCNEVQTNQAEYTTIIAPWNCADYTQWKNCDWDKTYPSVWFVRAGPFCKAMRHQCGISWYHCAGYATCQSYYFTTGPNIEAYGRWFPWCMVGKHNNNMSHYTEIDHCQNCHVWALLRMDYNPELPSMQNWIPSEAYPNLMHTGNNTAMCTATPWNPNLVILVCTSGYNLSINHIWHCMEATDIHDGIHTVYCWYPFCSFIDCMLHYHQPMEKWTQHRGNCNTLNREHRNMSLNHMRDITYVVRWNFNFRLPAACCFGHQYTAVWLYYENSVYKTRQKGNSHGIRDCQSAYMPVIVGQMNRDILCDKRIKIKNYGVPNCRSEFYNAAVIFFECELTPMLFLIVAQISYHECVSYIDNMFRIFGRFTFLYHDEMYQAHVMQSSELNAIDMTGNMSMDGIGYFHPWYAKYGNTVITHALEQTYFFGYYRISIYKSNRRGKNPTDDCDLPTWIIHISCVSSMRHVKDHYKLDEVHHNSIDVLMLSRLLKNVPQQGCQVKMTMMYPAYIAMFKREQWHWAHKDEECATQKKMAWAVGVQQIVWRKWYHIEWCVFFGLYKVWVLWSMFAHAAMACSNYENLEHAWWVGNMEMKHCMCYWQEASIWTRSIHYKMYFHGSGQKNVFQWPRETFGTPYLLYFNIRDTQEFHSPTPSLKGYHSWGVMHVGKMEVSDTGETIHMMMIFHGYGYLISWCITLWNNWPKRKCYITYKHSYHPRYWNKNHAVDLYCNECQGTSTPHSGTPSRHFLMNTCALNSE'

def protein_mass(seq):
  mass = 0
  for P in seq:
    mass += protein_mass_dict[P]
  return mass

protein_mass(protein)

# REVP: Locating Restriction Sites

Given: A DNA string of length at most 1 kbp in FASTA format.

Return: The position and length of every reverse palindrome in the string having length between 4 and 12. You may return these pairs in any order.

Rosalind_4756 = 
'GCGCTCCCGATGGCTGCTCTTGTAAAGTTCCACTCAGCATCTATCGTCATGCCCAAGCGATTGCCTGGACTTTCCTCGCACAGATATAGGCTTGTGGACATCGCTCGCTGCTAGGTTGAGTTTTGCCATCAGGGACTACGATTCTGTCCACAAGAATGAATGTACAGATCGGCTGGGCCGCAGTGCATCCACCTAAAAGGAAAGATGTGATACAGTTGCTCTAGAGCGCCGCGGTATCGTGTTTGAGCAGGTAGTCGCCATGTGTCAGGTCCAGCAGCGAGACACCATACCCCAACCACTCACGAGCACCTGCGGCTCATGATTTTCGAACACGCTGAAGACCTCAGGGGGCCGTAGGTAACGGATGGAAACGTCAAGCTCGGCCATAGGCGCGAACCGACCGCAATGCTAGCGCGCGCTAATTTACGGCGAGCACATGCCATCATACTTTCAGCACTCCTGTACGCCCTAACTAGGTGCCTAGTAATATCGATAAATAGCCGGCTCCCATCAGAAATGGCGAGCTAACTTATGAACACAGGGGACGTTACAATTGTCTTCAACACGTGAGCTGCCAGAAGACACTCATTCGTTCTTGGTTTTGATCGGTTAGGAACACCGTCCTAATTCGTCAACAGGGTAACATAGGCAGCAAAAGGAGATCATAGCTTACTGGTCCGTTGCAACCACGACTGTATCACACTGAATATTACAGTAGGCGTCTATCTACCGAACCATCATACGCTATTGGATCACAAGAAGGTCGAGAAAGGAAAACCCGTTACTCACGATCCGATGGGCCACATGTCCCCCGGCTACGCAACTTGCGC'
def rev_palindromes(seq):

  bp_dict = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}
  def rev_comp(dnaseq):
    rev_dna = dnaseq[::-1]
    comp = ''
    for nt in rev_dna: 
      comp += bp_dict[nt]
    return comp

  results = []

  for i in range(len(seq)):
    for length in range(4, 13):
      fragment = seq[i:i+length]
      if len(fragment) == length:
        if fragment == rev_comp(fragment):
          results.append((i+1, length))

  return results

rev_palindromes(Rosalind_4756)

# TRAN: Transitions and Tranversions 

Given: Two DNA strings s1 and s2 of equal length (at most 1 kbp).

Return: The transition/transversion ratio R(s1, s2). 

Rosalind_0209 = 'GCAACGCACAACGAAAACCCTTAGGGACTGGATTATTTCGTGATCGTTGTAGTTATTGGAAGTACGGGCATCAACCCAGTT'
Rosalind_2200 = 'TTATCTGACAAAGAAAGCCGTCAACGGCTGGATAATTTCGCGATCGTGCTGGTTACTGGCGGTACGAGTGTTCCTTTGGGT'

def tt_ratio(seq1, seq2):
  transition = {('A', 'G'), ('G', 'A'), ('C', 'T'), ('T', 'C')}
  ts = 0
  tv = 0

  for a, b in zip(seq1, seq2):
    if a != b:
      if (a, b) in transition:
        ts += 1
      else:
        tv += 1
  return ts/tv

tt_ratio(Rosalind_0209, Rosalind_2200)
